pkg/gadgets/tcpconnect/tracer/core/tracer.go

```diff
@@ -0,0 +1,248 @@
+
+const PERF_BUFFER_PAGES = 64
+
+type Tracer struct {
+	config        *tracer.Config
+	resolver      containercollection.ContainerResolver
+	eventCallback func(types.Event)
+	node          string
+
+	objs        tcpconnectObjects
+	v4EnterLink link.Link
+	v4ExitLink  link.Link
+	v6EnterLink link.Link
+	v6ExitLink  link.Link
+	reader      *perf.Reader
+}
+
+func NewTracer(config *tracer.Config, resolver containercollection.ContainerResolver,
+	eventCallback func(types.Event), node string) (*Tracer, error) {
+	t := &Tracer{config: config}
+
+	t.resolver = resolver
+	t.eventCallback = eventCallback
+	t.node = node
+
+	if err := t.start(); err != nil {
+		t.Stop()
+		return nil, err
+	}
+
+	return t, nil
+}
+
+func (t *Tracer) Stop() {
+	t.v4EnterLink = gadgets.CloseLink(t.v4EnterLink)
+	t.v4ExitLink = gadgets.CloseLink(t.v4ExitLink)
+	t.v6EnterLink = gadgets.CloseLink(t.v6EnterLink)
+	t.v6ExitLink = gadgets.CloseLink(t.v6ExitLink)
+
+	t.objs.Close()
+}
+
+func (t *Tracer) start() error {
+	var err error
+	spec, err := loadTcpconnect()
+	if err != nil {
+		return fmt.Errorf("Failed to load ebpf program: %w", err)
+	}
+
+	filter_by_mnt_ns := false
+
+	if t.config.MountnsMap != "" {
+		filter_by_mnt_ns = true
+		m := spec.Maps["mount_ns_set"]
+		m.Pinning = ebpf.PinByName
+		m.Name = filepath.Base(t.config.MountnsMap)
+	}
+
+	consts := map[string]interface{}{
+		"filter_by_mnt_ns": filter_by_mnt_ns,
+	}
+
+	if err := spec.RewriteConstants(consts); err != nil {
+		return fmt.Errorf("error RewriteConstants: %w", err)
+	}
+
+	opts := ebpf.CollectionOptions{
+		Maps: ebpf.MapOptions{
+			PinPath: filepath.Dir(t.config.MountnsMap),
+		},
+	}
+
+	if err := spec.LoadAndAssign(&t.objs, &opts); err != nil {
+		return fmt.Errorf("Failed to load ebpf program: %w", err)
+	}
+
+	t.v4EnterLink, err = link.Kprobe("tcp_v4_connect", t.objs.TcpV4Connect)
+	if err != nil {
+		return fmt.Errorf("error attaching program: %w", err)
+	}
+
+	t.v4ExitLink, err = link.Kretprobe("tcp_v4_connect", t.objs.TcpV4ConnectRet)
+	if err != nil {
+		return fmt.Errorf("error attaching program: %w", err)
+	}
+
+	t.v6EnterLink, err = link.Kprobe("tcp_v6_connect", t.objs.TcpV6Connect)
+	if err != nil {
+		return fmt.Errorf("error attaching program: %w", err)
+	}
+
+	t.v6ExitLink, err = link.Kretprobe("tcp_v6_connect", t.objs.TcpV6ConnectRet)
+	if err != nil {
+		return fmt.Errorf("error attaching program: %w", err)
+	}
+
+	reader, err := perf.NewReader(t.objs.tcpconnectMaps.Events, PERF_BUFFER_PAGES*os.Getpagesize())
+	if err != nil {
+		return fmt.Errorf("error creating perf ring buffer: %w", err)
+	}
+	t.reader = reader
+
+	go t.run()
+
+	return nil
+}
+
+// TODO: move to utils or improve it!
+// htons converts an unsigned short integer from host byte order to network byte order.
+func htons(i uint16) uint16 {
```

Comment:
Your function seems correct but you can also use `htons` from C?

Reply:
That's a good idea. Given that we're already cgo the overhead won't be that much.